internal/catalogmetadata/cache/cache_test.go

```diff
@@ -0,0 +1,269 @@
+package cache_test
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"net/http"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	catalogd "github.com/operator-framework/catalogd/api/core/v1alpha1"
+
+	"github.com/operator-framework/operator-controller/internal/catalogmetadata/cache"
+	"github.com/operator-framework/operator-controller/internal/catalogmetadata/client"
+)
+
+const (
+	package1 = `{
+		"schema": "olm.package",
+		"name": "fake1"
+	}`
+
+	bundle1 = `{
+		"schema": "olm.bundle",
+		"name": "fake1.v1.0.0",
+		"package": "fake1",
+		"image": "fake-image",
+		"properties": [
+			{
+				"type": "olm.package",
+				"value": {"packageName":"fake1","version":"1.0.0"}
+			}
+		]
+	}`
+
+	stableChannel = `{
+		"schema": "olm.channel",
+		"name": "stable",
+		"package": "fake1",
+		"entries": [
+			{
+				"name": "fake1.v1.0.0"
+			}
+		]
+	}`
+)
+
+func TestClient(t *testing.T) {
+	t.Run("Bundles", func(t *testing.T) {
+		type test struct {
+			name       string
+			catalog    *catalogd.Catalog
+			contents   []byte
+			wantErr    bool
+			tripper    *MockTripper
+			extraTests func(t *testing.T, c client.Fetcher, ctx context.Context, tst test)
+		}
+		for _, tt := range []test{
+			{
+				name: "valid non-cached fetch",
+				catalog: &catalogd.Catalog{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "test-catalog",
+					},
+					Status: catalogd.CatalogStatus{
+						ResolvedSource: &catalogd.CatalogSource{
+							Type: catalogd.SourceTypeImage,
+							Image: &catalogd.ImageSource{
+								Ref: "fake/catalog@sha256:fakesha",
+							},
+						},
+					},
+				},
+				contents:   []byte(strings.Join([]string{package1, bundle1, stableChannel}, "\n")),
+				tripper:    &MockTripper{},
+				extraTests: func(_ *testing.T, _ client.Fetcher, _ context.Context, _ test) {},
+			},
+			{
+				name: "valid cached fetch",
+				catalog: &catalogd.Catalog{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "test-catalog",
+					},
+					Status: catalogd.CatalogStatus{
+						ResolvedSource: &catalogd.CatalogSource{
+							Type: catalogd.SourceTypeImage,
+							Image: &catalogd.ImageSource{
+								Ref: "fake/catalog@sha256:fakesha",
+							},
+						},
+					},
+				},
+				contents: []byte(strings.Join([]string{package1, bundle1, stableChannel}, "\n")),
+				tripper:  &MockTripper{},
+				extraTests: func(t *testing.T, c client.Fetcher, ctx context.Context, tst test) {
+					rc, err := c.FetchCatalogContents(ctx, tst.catalog)
+					assert.NoError(t, err)
+					defer rc.Close()
+					data, err := io.ReadAll(rc)
+					assert.NoError(t, err)
+					assert.Equal(t, tst.contents, data)
```

Comment:
I can see how these extra tests provide coverage of the cache code, but unless I'm missing something, they don't actually verify whether cached data or fetched data is returned.

Reply:
You're correct on this one - I missed forcing an update to the contents returned by the mock http.RoundTripper. The test below verifies this by ensuring the contents returned by the function are the same as the contents that should be returned by the `http.RoundTripper`

Reply:
For more clarity, the logic behind this is:
- If tst.tripper.content is changed and the cache is not used the content that will be returned != tst.contents
- If tst.tripper.content is changed _and_ the status.resolvedSource.image.ref is changed it won't hit the cache and thus the response should be different than original contents and instead match what the tst.tripper is told to return.

Reply:
(I've updated the code a bit to try and make this more clear when reading it)